zkevm-circuits/src/evm_circuit/execution/call.rs

```diff
@@ -0,0 +1,718 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{N_BYTES_ACCOUNT_ADDRESS, N_BYTES_GAS, N_BYTES_MEMORY_WORD_SIZE},
+        step::ExecutionState,
+        table::{AccountFieldTag, CallContextFieldTag, FixedTableTag, Lookup},
+        util::{
+            common_gadget::TransferGadget,
+            constraint_builder::{
+                ConstraintBuilder, ReversionInfo, StepStateTransition,
+                Transition::{Delta, To},
+            },
+            from_bytes,
+            math_gadget::{
+                BatchedIsZeroGadget, ConstantDivisionGadget, IsEqualGadget, IsZeroGadget,
+                MinMaxGadget,
+            },
+            memory_gadget::{MemoryAddressGadget, MemoryExpansionGadget},
+            select, sum, Cell, Word,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use eth_types::{
+    evm_types::{GasCost, GAS_STIPEND_CALL_WITH_VALUE},
+    Field, ToLittleEndian, ToScalar, EMPTY_HASH_LE,
+};
+use halo2_proofs::{circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct CallGadget<F> {
+    opcode: Cell<F>,
+    tx_id: Cell<F>,
+    reversion_info: ReversionInfo<F>,
+    caller_address: Cell<F>,
+    is_static: Cell<F>,
+    depth: Cell<F>,
+    gas: Word<F>,
+    callee_address: Word<F>,
+    value: Word<F>,
+    is_success: Cell<F>,
+    gas_is_u64: IsZeroGadget<F>,
+    is_warm_access: Cell<F>,
+    is_warm_access_prev: Cell<F>,
+    callee_reversion_info: ReversionInfo<F>,
+    value_is_zero: IsZeroGadget<F>,
+    cd_address: MemoryAddressGadget<F>,
+    rd_address: MemoryAddressGadget<F>,
+    memory_expansion: MemoryExpansionGadget<F, 2, N_BYTES_MEMORY_WORD_SIZE>,
+    transfer: TransferGadget<F>,
+    callee_nonce: Cell<F>,
+    callee_code_hash: Cell<F>,
+    is_account_empty: BatchedIsZeroGadget<F, 2>,
+    is_empty_code_hash: IsEqualGadget<F>,
+    one_64th_gas: ConstantDivisionGadget<F, N_BYTES_GAS>,
+    capped_callee_gas_left: MinMaxGadget<F, N_BYTES_GAS>,
+}
+
+impl<F: Field> ExecutionGadget<F> for CallGadget<F> {
+    const NAME: &'static str = "CALL";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CALL;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        cb.opcode_lookup(opcode.expr(), 1.expr());
+
+        // We do the `ResponsibleOpcode` lookup explicitly here because we're not using
+        // the `SameContextGadget` for `CALL`.
+        cb.add_lookup(
+            "Responsible opcode lookup",
+            Lookup::Fixed {
+                tag: FixedTableTag::ResponsibleOpcode.expr(),
+                values: [
+                    cb.execution_state().as_u64().expr(),
+                    opcode.expr(),
+                    0.expr(),
+                ],
+            },
+        );
+
+        let gas_word = cb.query_word();
+        let callee_address_word = cb.query_word();
```

Comment:
Aren't we suposed to `query_rlc` here??? I would like to understand why we aren't :) I'm sure I'm missing something.

Reply:
Oh nevermind. It's because the field is not represented as bytes and therefore we don't need to do so?
anyway I wonder what happens if we go over the field size.

Reply:
Yeah it's RLC here. `query_word` is just `query_rlc` but with fixed 32 bytes.

Reply:
Ohh that's right! Perfect!